### ISSUE TITLE
syntax highlighting adds space in code

### ISSUE THREAD

**Issue by schplurtz** [dokuwiki/dokuwiki#2614]

**Comment by schplurtz** [comment]
oops. Forgot the description (github should not allow this ;))

empty lines in `<code bash>` are turned into "one space line" when syntax highlighting is requested.

The following wiki code, where the line between `aaaa` and `bbbb` is empty is rendered with an additional space char. It's easy to show the problem just by selecting text, as is shown on the screenshot below (screenshot made on dokuwiki.org)
```
<code bash>
aaaa a=42

bbbb b=56
</code>
```
-------------

<img width="709" alt="image" src="https://user-images.githubusercontent.com/1008292/49338925-9a712b00-f629-11e8-8ce6-c445d1cfa4b2.png">

**Comment by Michaelsy** [comment]
> oops. Forgot the description (github should not allow this ;))

You can still edit the postings after submitting.

**Comment by Klap-in** [comment]
DokuWiki uses the Geshi library for highlighting: https://github.com/splitbrain/dokuwiki/tree/master/vendor/geshi/geshi
Testing at their site, without the processing eventually performed by DokuWiki, shows that Geshi already adds a space at that line (more precise: a non-breakable space). 
So my guess is that developers of Geshi has done this by purpose.

My guess is that it prevents empty html elements.

related discussions:
https://www.drupal.org/project/geshifilter/issues/248922
https://stackoverflow.com/questions/14848326/empty-paragraph-tags-should-i-allow-in-html-editor-or-not

**Comment by schplurtz** [comment]
Yes it changes the meaning of my code. I need some data organised in stanza. for example for ldapadd.
```
ldapadd -foobar <<EOH
dn: foo=bar,ou=baz,dc=example,dc=com
objectClass: schproutz
foo: bar

dn: foo=qux,ou=baz,dc=example,dc=com
objectClass: schproutz
foo: qux
EOH
```
With a non breakable space, You get only one stanza, and if you copy/paste the code, it fails miserably.

Going to read your links

**Comment by phy25** [comment]
In DokuWiki's usage (`ol>li>div`/`pre`) where `p` is not involved, I don't think anything is required inside the per-line element to display a line. However the behavior in GeSHi seems intended: https://github.com/GeSHi/geshi-1.0/commit/700e7946f4dcd2c258d3a261187a88687bb1325b (though I am not sure why lines can be collapsed in this case)

I would recommend using [Downloadable Code Blocks](https://www.dokuwiki.org/wiki:syntax#downloadable_code_blocks) to workaround this.

**Comment by schplurtz** [comment]
Yes this seems intended...

Downloadable code blocks are a solution but compared to copy and paste, there are so many manipulations needed, just to get the code block content into the clipboard.

If people are looking for another solution, I recently published [copy2clipboard](https://www.dokuwiki.org/plugin:copy2clipboard), a JS plugin that adds copy to clipboard button to any code block in the page. This plugin removes on purpose nobreak spaces when they are the only char on a line.

**Comment by schplurtz** [comment]
I'm not sure they had the correct reason at drupal. Indentation is based on nbsp in geshi when not using `<pre>` prefix, as one can read in the doc you pointed. 

> if you use `<pre>` the whitespace you use will appear pretty much exactly how it is in the source, and what’s more GeSHi won’t have to add a whole lot of `<br />`’s and non-breaking spaces (`&nbsp;`) to your code to indent it.

But yes this has been a problem since that time, @phy25  even found a commit that dates back to 2006 .

DW is facing the exact same situation, geshi highlighted code with `pre` gets `&nbsp;` on blank lines. Drupal OP solved this in his/her instance by writing a filter to remove the `&nbsp;`

**Comment by schplurtz** [comment]
When you say

> (btw, I don’t expect work on this feature from the side of DokuWiki. So, for more progress, you have to volunteer yourself or find someone who would work now on it for you.)

Do you mean that if I submitted a PR (with tests) for this, it could eventually be accepted in DW ? Not that I'm going to right now, but just to know if fixing upstream in DW is accepted practice.

**Comment by splitbrain** [comment]
A proper fix in GesHi would be preferred. We're currently using our own fork, because upstream is slow to accept and release new fixes, but we try to stay as close as possible to them. So, if you want to tackle this problem, please submit a PR against upstream and we will pull it into our fork.

**Comment by Quentin-ctrL** [comment]
Hello,

Any news on this ?
Is there any workaround if there is no fix yet ?

**Comment by Klap-in** [comment]
No, not from the side of DokuWiki. The best is to go upstream and report it (if not yet done) in the GesHi project, and help there if possible.
